Deprecated some functions, improved method documentation and test for isValidUrl

marcosgdf opened the pull request.

marcosgdf committed:
Deprecated some functions and improved method documentation

marcosgdf commented on htdocs/core/lib/functions.lib.php:
```diff
@@ -3631,20 +3631,19 @@ function dol_string_nohtmltag($StringHtml,$removelinefeed=1,$pagecodeto='UTF-8')
-		// We use @ to avoid warning on PHP4 that does not support entity encoding from UTF8;
-		if (version_compare(PHP_VERSION, '5.3.0') < 0) return @nl2br($stringtoencode);
-		else return @nl2br($stringtoencode,$forxml);
+		if (version_compare(PHP_VERSION, '5.3.0') < 0) return nl2br($stringtoencode);
```
I removed @ because it was only introduced for PHP 4

eldy commented:
It seems you lost paramaeter ",$forxml" into patch.

marcosgdf commented:
Nope. $forxml is only available since PHP 5.3.0, that's why the comparison is made. I do not know what is the min PHP version required right now.

Also... it would be interesting to know if giving the 2nd parameter in <5.3.0 will break or not the script

eldy commented:
Minimum versions are defined here:
http://wiki.dolibarr.org/index.php/Prerequisite
So we must keep the test but we can remove the @

rdoursenaud commented:
The README says PHP5.3+ you bumped it
at 7a5030826b9667b9219f18345be5833cea909db1 so I think the wiki is out of
sync.

2014-07-16 20:46 GMT+02:00 Laurent Destailleur notifications@github.com:

> Minimum versions are defined here:
> http://wiki.dolibarr.org/index.php/Prerequisite
> So we must keep the test but we can remove the @
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Dolibarr/dolibarr/pull/1748#issuecomment-49209341.

## 

_Raphaël Doursenaud_
Directeur technique (CTO)
Expert certifié en déploiement Google Apps
https://gpcsolutions.fr/raphael-doursenaud-google-apps-certified-deployment-specialist
+33 (0)5 35 53 97 13 - +33 (0)6 68 48 20 10

## 

 http://gpcsolutions.fr
http://gpcsolutions.fr
Technopole Hélioparc
2 avenue du Président Pierre Angot
64053 PAU CEDEX 9
SARL GPC.solutions au capital de 7 500 € - R.C.S. PAU 528 995 921
https://www.google.com/a/partnersearch/#partner?partner_id=46687933_a0n60000000sqpWAAQ 
http://wiki.dolibarr.org/index.php/Dolibarr_suppliers_France#GPC.solutions

marcosgdf committed:
Removed Dolibarr implementation of json_encode and json_decode, and deprecated dol_json_encode

marcosgdf committed:
Optimized isValidEmail function and load json.lib.php always as it contains common Dolibarr functions

marcosgdf committed:
Deprecated function dol_json_decode and replaced with native behaviour

marcosgdf committed:
Copyright

marcosgdf committed:
Corrected dol_microtime_float refactor

marcosgdf committed:
Added a test for isValidUrl function

marcosgdf committed:
Testing dol_json_encode is not interesting now as it uses a core function

marcosgdf committed:
Removed empty test

marcosgdf committed:
Corrected testIsValidUrl

marcosgdf committed:
Removed PHP 5.3 comparison in dol_nl2br function

marcosgdf committed:
Created test for dol_nl2br

marcosgdf committed:
Deprecated getStaticMember

marcosgdf committed:
Removed PHP 5.3 comparison in ImportCsv

marcosgdf committed:
Added missing return null for getStaticMember

eldy commented:
The goal of library file '/core/lib/json.lib.php' is to provide an implementation of json_encode when it is not available.

So the
-if (! function_exists('json_encode'))
-{

include_once DOL_DOCUMENT_ROOT .'/core/lib/json.lib.php'; -}
must not be replace with include, but completely removed, like the file json.lib.php.

Also, even with php 5.3, json is not always available as it is an external module and not "by default" in php, so i think we must keep current code. Above all, I experienced different implementation of json, so u prefere keeping unit test on json because it was the only solution we had to detect difference of implementation and make an alert something is wrong with php before loosing time with very strang behaviour of code.

So can you provide another PR with just changes done onto "@" and keep all json functions ?

marcosgdf commented:
According to PHP documentation, JSON is bundled and compiled by default since PHP 5.2. So, unless you explicitly remove JSON module from PHP, it will be bundled. (http://php.net/manual/en/json.installation.php)

If you still want to have the reimplementation of json functions, tell me and I will remove those changes.

Ok, I will restore the JSON tests. 

About the inclusion of json.lib.php without checking if json functions are available it's because they are creating functions that are not JSON-centric related, like utf162utf8 and utf82utf16 that developers might use.

marcosgdf committed:
Revert "Removed empty test"

This reverts commit cf1a2c44a359291288d7cd8f49b86ca4bb36d26a.

eldy commented:
Sources are embended but into a nexternal module. In linux distributions they are packaged into external package (for example with ubuntu/debian meaning it is in packge php5-json). This means it is not always availble on some web host provider. Or some light apachhe installation. So i prefer keep the old implementation and use new one only if available.

marcosgdf commented:
Yeah but according to PHP documentation, php5 depends on php5-json. Here is an example http://packages.ubuntu.com/trusty/php5-common

Will restore it though.

marcosgdf committed:
Restored JSON reimplementation

marcosgdf committed:
Revert "Testing dol_json_encode is not interesting now as it uses a core function"

marcosgdf committed:
Removed duplicated requires.
Json.lib.php is included with functions.lib.php

marcosgdf committed:
Replaced usage of dol_json_encode and dol_json_decode with json_encode and json_decode. If json module is not installed, it will refer to dol_json_encode/decode

marcosgdf commented:
I have now:
- Restored Dolibarr's JSON implementation
- Removed json.lib.php includes from all over Dolibarr except Stats class, because it is now included in functions.lib.php
- Replaced `dol_json_encode` and `dol_json_decode` function calls with `json_encode` and `json_decode` because if JSON module is not available, it will redir to dol_json_decode/encode.
